kem/frodo/frodo640shake/frodo.go

```diff
@@ -0,0 +1,628 @@
+// Package frodo640shake implements the IND-CCA2 secure key encapsulation mechanism
+// FrodoKEM-640-SHAKE as submitted to round 3 of the NIST PQC competition and
+// described in
+//
+// https://frodokem.org/files/FrodoKEM-specification-20210604.pdf
+package frodo640shake
+
+import (
+	"github.com/cloudflare/circl/kem"
+	"github.com/cloudflare/circl/xof"
+
+	"bytes"
+	cryptoRand "crypto/rand"
+	"io"
+)
+
+const (
+	paramN             = 640
+	paramNbar          = 8
+	logQ               = 15
+	seedASize          = 16
+	pkHashSize         = 16
+	extractedBits      = 2
+	messageSize        = 16
+	matrixBpPackedSize = (logQ * (paramN * paramNbar)) / 8
+)
+
+const (
+	// Size of seed for NewKeyFromSeed
+	KeySeedSize = 2*SharedKeySize + seedASize
+
+	// Size of seed for EncapsulateTo.
+	EncapsulationSeedSize = 16
+
+	// Size of the established shared key.
+	SharedKeySize = 16
+
+	// Size of the encapsulated shared key.
+	CiphertextSize = 9720
+
+	// Size of a packed public key.
+	PublicKeySize = 9616
+
+	// Size of a packed private key.
+	PrivateKeySize = 19888
+)
+
+// Type of a FrodoKEM-640-SHAKE public key
+type PublicKey struct {
+	seedA   [seedASize]byte
+	matrixB [paramN * paramNbar]uint16
+}
+
+// Type of a FrodoKEM-640-SHAKE private key
+type PrivateKey struct {
+	hashInputIfDecapsFail [SharedKeySize]byte
+	pk                    *PublicKey
+	matrixS               [paramN * paramNbar]uint16
+	hpk                   [pkHashSize]byte // H(packed(pk))
+}
+
+// NewKeyFromSeed derives a public/private keypair deterministically
+// from the given seed.
+//
+// Panics if seed is not of length KeySeedSize.
+func newKeyFromSeed(seed []byte) (*PublicKey, *PrivateKey, error) {
+	if len(seed) != KeySeedSize {
+		panic("seed must be of length KeySeedSize")
+	}
+
+	var sk PrivateKey
+	var pk PublicKey
+
+	var shakeInputForSE [1 + SharedKeySize]byte
+
+	var SE [2 * paramN * paramNbar]uint16
+	var byteSE [2 * len(SE)]byte
+	S := SE[0 : paramN*paramNbar]
+	E := SE[paramN*paramNbar : 2*paramN*paramNbar]
+
+	var A [paramN * paramN]uint16
+
+	// Generate the secret value s, and the seed for S, E, and A. Add seedA to the public key
+	shake128 := xof.SHAKE128.New()
```

Comment:
Instead of using the interface, it might be faster to use the shake128 package directly (as is done in Kyber for instance) as that can help prevent allocations due to making it easier for Go's escape analysis.